### PR TITLE
Extend SeccompCondition with arg len parameter

### PR DESCRIPTION
**Pull request by serban300**
Signed-off-by: Serban Iorga <seriorga@amazon.com>

Issue #, if available: #1206 

Description of changes: 

Some libc implementations (for example musl) perform sign extension when
passing an int argument to a linux system call. This can result in
incorrect behavior when a bpf rule treats any syscall arguments as
a 64 bits variable.

We define a way to compare only the least significant 32 bits of a
syscall argument in order to deal with this corner case.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: vmm/src/default_syscalls/filters.rs [hunk 1]
```diff
@@ -40,8 +40,8 @@ pub fn default_filter() -> Result<SeccompFilter, Error> {
             allow_syscall_if(
                 libc::SYS_epoll_ctl,
                 or![
-                    and![Cond::new(1, Eq, super::EPOLL_CTL_ADD)?],
-                    and![Cond::new(1, Eq, super::EPOLL_CTL_DEL)?],
+                    and![Cond::new(1, ArgLen::DWORD, Eq, super::EPOLL_CTL_ADD)?],
```
Comment (acatangiu) [inline vmm/src/default_syscalls/filters.rs]:
```suggestion
                    Cond::new(1, ArgLen::DWORD, Eq, super::EPOLL_CTL_ADD)?,
```

File: vmm/src/default_syscalls/filters.rs [hunk 2]
```diff
@@ -40,8 +40,8 @@ pub fn default_filter() -> Result<SeccompFilter, Error> {
...
+                    and![Cond::new(1, ArgLen::DWORD, Eq, super::EPOLL_CTL_DEL)?],
```
Comment (acatangiu) [inline vmm/src/default_syscalls/filters.rs]:
```suggestion
                    Cond::new(1, ArgLen::DWORD, Eq, super::EPOLL_CTL_DEL)?,
```

File: vmm/src/default_syscalls/filters.rs [hunk 3]
```diff
@@ -62,11 +62,21 @@ pub fn default_filter() -> Result<SeccompFilter, Error> {
             allow_syscall_if(
                 libc::SYS_futex,
                 or![
-                    and![Cond::new(1, Eq, super::FUTEX_WAIT_PRIVATE)?],
-                    and![Cond::new(1, Eq, super::FUTEX_WAKE_PRIVATE)?],
-                    and![Cond::new(1, Eq, super::FUTEX_REQUEUE_PRIVATE)?],
+                    and![Cond::new(1, ArgLen::DWORD, Eq, super::FUTEX_WAIT_PRIVATE)?],
```
Comment (acatangiu) [inline vmm/src/default_syscalls/filters.rs]: Same here, I don't think all of these conditions need to be wrapped in `and!`s.

**Comment by serban300** [comment]
> It'd be nice to address all the extraneous `and!`s throughout the rules that you updated, but I realise it's not necessary in this PR so your choice.

I just tried it. It doesn't work without the `and!` . The error is:

```
expected struct `seccomp::SeccompRule`, found struct `seccomp::SeccompCondition
```

Probably we could add a new macro `allow!` and replace `and!` with `allow!` in some cases for clarity. I don't know. It's debatable. But I wouldn't do it in this PR since it already touched a lot of code.

File: seccomp/src/lib.rs [hunk 4]
```diff
@@ -1198,7 +1756,68 @@ mod tests {
...
+        // Compares translated filter with hardcoded BPF program.
+        {
+            let mut empty_rule_map = BTreeMap::new();
+            empty_rule_map.insert(1, vec![]);
+            assert!(SeccompFilter::new(empty_rule_map, SeccompAction::Allow).is_err());
+        }
+
+        let filter = create_test_bpf_filter(ArgLen::DWORD);
+
+        let instructions = vec![
```
Comment (andreeaflorescu) [inline seccomp/src/lib.rs]: How did you obtain the following hardcoded instructions? Please add a comment with the steps so people that want to change this in the future know how to.

File: seccomp/src/lib.rs [hunk 5]
```diff
@@ -376,10 +376,21 @@ pub enum SeccompCmpOp {
     Ne,
 }
 
+/// Secoomp argument length.
```
[the remark this answers is not in the document]
Comment (serban300) [reply seccomp/src/lib.rs]: Fixed

File: seccomp/src/lib.rs [hunk 6]
```diff
@@ -376,10 +376,21 @@ pub enum SeccompCmpOp {
...
+    DWORD,
+    /// Argument length is 8 bytes.
+    QWORD,
+}
+
 /// Condition that syscall must match in order to satisfy a rule.
 pub struct SeccompCondition {
     /// Number of the argument value that is to be compared.
     arg_number: u8,
+    /// Length of the argument that is to be compared
```
[the remark this answers is not in the document]
Comment (serban300) [reply seccomp/src/lib.rs]: Fixed

File: seccomp/src/lib.rs (hunk 4 above)
Comment (serban300) [reply seccomp/src/lib.rs]: I did it manually. I started from the instructions for `test_filter_bpf_output_qword` which were already present and modified them accordingly. I don't think it's worth a comment.

Comment (acatangiu) [reply seccomp/src/lib.rs]: You simply need to learn the voodoo incantation 😆

Comment (andreeaflorescu) [reply seccomp/src/lib.rs]: Ok. Hope you'll modify them the next time cause I ain't got any idea how to do that.

**Pull request merged**